File: tools/pairing-store-server.mjs
#!/usr/bin/env node
// Pairing-session store for /iptv, to run on your own VPS instead of a hosted
// Redis. It speaks the small subset of the Upstash REST protocol that
// lib/iptv-store.ts actually uses, so the website needs no code change: point
// UPSTASH_REDIS_REST_URL at this service and UPSTASH_REDIS_REST_TOKEN at the
// token below.
//
// Sessions live in this process's memory with a TTL. That is correct here, and
// not the compromise it would be on serverless, precisely because this is one
// long-lived process: the phone's write and the TV's read always meet.
//
//   Run:   PAIRING_TOKEN=<long-random> node pairing-store-server.mjs
//   Env:   PAIRING_TOKEN (required)  PORT (default 8787)  HOST (default 127.0.0.1)
//
// Terminate TLS in front of it (nginx/Caddy) and bind to localhost: the token
// is the only authentication, and payloads hold IPTV credentials.

import http from "node:http"
import { timingSafeEqual } from "node:crypto"

const TOKEN = process.env.PAIRING_TOKEN
const PORT = Number(process.env.PORT ?? 8787)
const HOST = process.env.HOST ?? "127.0.0.1"
const MAX_BODY_BYTES = 256 * 1024

if (!TOKEN || TOKEN.length < 24) {
  console.error("PAIRING_TOKEN is required and must be at least 24 characters.")
  process.exit(1)
}

/** key -> { value, expiresAt } */
const store = new Map()

setInterval(() => {
  const now = Date.now()
  for (const [key, entry] of store) if (entry.expiresAt <= now) store.delete(key)
}, 30_000).unref()

function authorised(header) {
  const supplied = (header ?? "").replace(/^Bearer /, "")
  const a = Buffer.from(supplied)
  const b = Buffer.from(TOKEN)
  return a.length === b.length && timingSafeEqual(a, b)
}

function run(command) {
  const [verb, key, ...rest] = command
  switch (String(verb).toUpperCase()) {
    case "SET": {
      const value = rest[0]
      const exIndex = rest.findIndex((part) => String(part).toUpperCase() === "EX")
      const ttl = exIndex >= 0 ? Number(rest[exIndex + 1]) : 600
      if (typeof value !== "string" || !Number.isFinite(ttl) || ttl <= 0) return { error: "ERR bad SET" }
      store.set(key, { value, expiresAt: Date.now() + ttl * 1000 })
      return { result: "OK" }
    }
    case "GETDEL": {
      const entry = store.get(key)
      if (!entry) return { result: null }
      store.delete(key)
      return { result: entry.expiresAt > Date.now() ? entry.value : null }
    }
    case "DEL": {
      return { result: store.delete(key) ? 1 : 0 }
    }
    case "PING":
      return { result: "PONG" }
    default:
      return { error: `ERR unsupported command` }
  }
}

const server = http.createServer((req, res) => {
  const send = (status, body) => {
    const text = JSON.stringify(body)
    res.writeHead(status, { "Content-Type": "application/json", "Cache-Control": "no-store" })
    res.end(text)
  }

  if (req.method !== "POST") return send(405, { error: "ERR method" })
  if (!authorised(req.headers.authorization)) return send(401, { error: "NOAUTH" })

  let size = 0
  const chunks = []
  req.on("data", (chunk) => {
    size += chunk.length
    if (size > MAX_BODY_BYTES) {
      send(413, { error: "ERR too large" })
      req.destroy()
      return
    }
    chunks.push(chunk)
  })
  req.on("end", () => {
    if (res.writableEnded) return
    let command
    try {
      command = JSON.parse(Buffer.concat(chunks).toString("utf8"))
    } catch {
      return send(400, { error: "ERR parse" })
    }
    if (!Array.isArray(command) || command.length === 0) return send(400, { error: "ERR shape" })
    // Values are IPTV credentials: log the verb, never the key or the value.
    const outcome = run(command)
    if (outcome.error) console.error(`[pairing-store] ${String(command[0]).toUpperCase()} rejected`)
    send(200, outcome)
  })
})

server.listen(PORT, HOST, () => {
  console.log(`[pairing-store] listening on ${HOST}:${PORT}`)
})

File: tools/pairing_store_server.py
#!/usr/bin/env python3
"""Pairing-session store for /iptv — Python port of pairing-store-server.mjs.

Speaks the small subset of the Upstash REST protocol that lib/iptv-store.ts
uses, so the website needs no code change: point UPSTASH_REDIS_REST_URL at this
service and UPSTASH_REDIS_REST_TOKEN at the token it is started with.

Standard library only, so it runs on a box that has Python but no Node and
needs no packages installed. Sessions live in this process's memory with a TTL,
which is correct here — unlike on serverless — because a single long-lived
process handles both the phone's write and the TV's read.

  Run:  PAIRING_TOKEN=<long-random> python3 pairing_store_server.py
  Env:  PAIRING_TOKEN (required)  PORT (default 8787)  HOST (default 127.0.0.1)

Terminate TLS in front of it and bind to localhost: the token is the only
authentication, and payloads hold IPTV credentials.
"""

import hmac
import json
import os
import sys
import threading
import time
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer

TOKEN = os.environ.get("PAIRING_TOKEN", "")
PORT = int(os.environ.get("PORT", "8787"))
HOST = os.environ.get("HOST", "127.0.0.1")
MAX_BODY_BYTES = 256 * 1024
SWEEP_INTERVAL = 30.0

if len(TOKEN) < 24:
    sys.stderr.write("PAIRING_TOKEN is required and must be at least 24 characters.\n")
    raise SystemExit(1)

# key -> (value, expires_at). The lock is what makes GETDEL atomic across
# threads: two TVs polling the same code must never both be handed the payload.
_store: dict[str, tuple[str, float]] = {}
_lock = threading.Lock()


def _sweep() -> None:
    while True:
        time.sleep(SWEEP_INTERVAL)
        now = time.monotonic()
        with _lock:
            for key in [k for k, (_, exp) in _store.items() if exp <= now]:
                del _store[key]


def run_command(command: list) -> dict:
    verb = str(command[0]).upper() if command else ""
    key = str(command[1]) if len(command) > 1 else ""

    if verb == "SET":
        value = command[2] if len(command) > 2 else None
        ttl = 600
        rest = [str(part).upper() for part in command[3:]]
        if "EX" in rest:
            try:
                ttl = int(command[3 + rest.index("EX") + 1])
            except (IndexError, TypeError, ValueError):
                return {"error": "ERR bad SET"}
        if not isinstance(value, str) or ttl <= 0:
            return {"error": "ERR bad SET"}
        with _lock:
            _store[key] = (value, time.monotonic() + ttl)
        return {"result": "OK"}

    if verb == "GETDEL":
        with _lock:
            entry = _store.pop(key, None)
        if entry is None:
            return {"result": None}
        value, expires_at = entry
        return {"result": value if expires_at > time.monotonic() else None}

    if verb == "DEL":
        with _lock:
            return {"result": 1 if _store.pop(key, None) is not None else 0}

    if verb == "PING":
        return {"result": "PONG"}

    return {"error": "ERR unsupported command"}


class Handler(BaseHTTPRequestHandler):
    protocol_version = "HTTP/1.1"
    server_version = "pairing-store"
    sys_version = ""

    def _send(self, status: int, body: dict) -> None:
        payload = json.dumps(body).encode("utf-8")
        self.send_response(status)
        self.send_header("Content-Type", "application/json")
        self.send_header("Content-Length", str(len(payload)))
        self.send_header("Cache-Control", "no-store")
        self.end_headers()
        self.wfile.write(payload)

    def _authorised(self) -> bool:
        supplied = self.headers.get("Authorization", "")
        if supplied.startswith("Bearer "):
            supplied = supplied[len("Bearer "):]
        return hmac.compare_digest(supplied, TOKEN)

    def do_POST(self) -> None:  # noqa: N802 - name fixed by BaseHTTPRequestHandler
        if not self._authorised():
            return self._send(401, {"error": "NOAUTH"})

        try:
            length = int(self.headers.get("Content-Length", "0"))
        except ValueError:
            return self._send(400, {"error": "ERR length"})
        if length > MAX_BODY_BYTES:
            return self._send(413, {"error": "ERR too large"})

        try:
            command = json.loads(self.rfile.read(length).decode("utf-8"))
        except (ValueError, UnicodeDecodeError):
            return self._send(400, {"error": "ERR parse"})
        if not isinstance(command, list) or not command:
            return self._send(400, {"error": "ERR shape"})

        outcome = run_command(command)
        if "error" in outcome:
            # Values are IPTV credentials: log the verb, never key or value.
            sys.stderr.write(f"[pairing-store] {str(command[0]).upper()} rejected\n")
        self._send(200 if "error" not in outcome else 200, outcome)

    def do_GET(self) -> None:  # noqa: N802
        self._send(405, {"error": "ERR method"})

    def log_message(self, *args) -> None:
        """Silence the access log: request lines would carry no secrets today,
        but the default handler is one refactor away from logging a body."""


if __name__ == "__main__":
    threading.Thread(target=_sweep, daemon=True).start()
    server = ThreadingHTTPServer((HOST, PORT), Handler)
    sys.stderr.write(f"[pairing-store] listening on {HOST}:{PORT}\n")
    sys.stderr.flush()
    server.serve_forever()
